x-pack/plugins/security_solution/public/management/pages/policy/store/policy_details/middleware/policy_trusted_apps_middleware.ts

```diff
@@ -5,12 +5,200 @@
+    ];
+
+    if (filter) {
+      const filterKuery = parseQueryFilterToKQL(filter, SEARCHABLE_FIELDS) || undefined;
+      if (filterKuery) kuery.push(filterKuery);
+    }
+
+    const trustedApps = await trustedAppsService.getTrustedAppsList({
+      page: 1,
+      per_page: 100,
+      kuery: kuery.join(' AND '),
+    });
+
+    store.dispatch({
+      type: 'policyArtifactsAvailableListPageDataChanged',
+      payload: createLoadedResourceState({
+        items: trustedApps.data,
+        pageIndex: 1,
+        pageSize: 100,
+        totalItemsCount: trustedApps.total,
+        timestamp: Date.now(),
+        filter: filter || '',
+        excludedPolicies: '',
+        includedPolicies: policyId,
+      }),
+    });
+
+    if (isEmpty(trustedApps.data)) {
+      checkIfThereAreAvailableTrustedApps(store, trustedAppsService);
+    }
+  } catch (err) {
+    store.dispatch({
+      type: 'policyArtifactsAvailableListPageDataChanged',
+      // Ignore will be fixed with when AsyncResourceState is refactored (#830)
+      // @ts-ignore
+      payload: createFailedResourceState(err.body ?? err),
+    });
+  }
+};
+
+const updateTrustedApps = async (
+  store: ImmutableMiddlewareAPI<PolicyDetailsState, PolicyDetailsAction>,
+  trustedAppsService: TrustedAppsService,
+  trustedAppsIds: ImmutableArray<string>
+) => {
+  const state = store.getState();
+  const policyId = policyIdFromParams(state);
+  const availavleArtifacts = getAvailableArtifactsList(state);
+
+  if (!availavleArtifacts || !availavleArtifacts.items.length) {
+    return;
+  }
+
+  store.dispatch({
+    type: 'policyArtifactsUpdateTrustedAppsChanged',
+    // Ignore will be fixed with when AsyncResourceState is refactored (#830)
+    // @ts-ignore
+    payload: createLoadingResourceState({ previousState: createUninitialisedResourceState() }),
+  });
+
+  try {
+    const trustedAppsUpdateActions = [];
+
+    for (const entry of availavleArtifacts.items) {
+      if (trustedAppsIds.includes(entry.id)) {
+        const policies = entry.effectScope.type === 'policy' ? entry.effectScope.policies : [];
+        const trustedApp = trustedAppsService.updateTrustedApp({ id: entry.id }, {
+          effectScope: { type: 'policy', policies: [...policies, policyId] },
+          name: entry.name,
+          entries: entry.entries,
+          os: entry.os,
+          description: entry.description,
+          version: entry.version,
+        } as PostTrustedAppCreateRequest);
+        trustedAppsUpdateActions.push(trustedApp);
+      }
+    }
+
+    const updatedTrustedApps = await Promise.all(trustedAppsUpdateActions);
+    store.dispatch({
+      type: 'policyArtifactsUpdateTrustedAppsChanged',
+      payload: createLoadedResourceState(updatedTrustedApps),
+    });
+  } catch (err) {
+    store.dispatch({
+      type: 'policyArtifactsUpdateTrustedAppsChanged',
+      // Ignore will be fixed with when AsyncResourceState is refactored (#830)
+      // @ts-ignore
```

Comment:
FYI - I think this one can be fixed by setting the type on the `createFailedResourceState()` (its a generic function)